### PR TITLE
provider/vsphere: cache VMDKs in datastore

### PR DESCRIPTION
**Pull request by axw**
## Description of change

Cache VMDKs in a controller-specific location
in the datastore, and use that when creating
VMs. This serves two purposes:

 - to avoid downloading and uploading for each
   new virtual machine
 - to avoid having to upload to the hosts, as
   we were previously doing (uploading to the
   NFC lease location)

We now also avoid downloading the OVA to disk,
instead piping the VMDK contents to the datastore
upload.

The VMDK we upload to the datastore is not stored
in the virtual disk format, so to create a functioning
VM we must:
 1. create a temporary VM with the VMDK attached
 2. clone that VM to create a copy of the VMDK in
   the virtual disk format

We continue to use OVF/ImportVApp for the purpose
of creating the initial VM, which keeps things the
same between vCenter and ESXi. Instead of pulling
the OVF out of the OVA file, we now store an OVF
definition in the source tree and compile it in. This
means we don't have to read the OVA file at all if
the VMDK is already cached.

This PR is based on top of https://github.com/juju/juju/pull/7855, so you can ignore the first commit.

## QA steps

1. juju bootstrap vsphere
2. juju add-machine --constraints root-disk=20G
3. juju add-machine --series=trusty
(ensure all machines come up OK, and check in debug-log that the machine added in step 2 did not incur an upload; ssh to machine from step 2 to check its root disk size is 20GiB)
4. check datastore, it should have VMDKs cached in juju-vmdks/<controller-uuid>/{xenial,trusty}.
5. delete the cache from datastore
6. juju add-machine
(ensure machine comes up OK)
7 juju destroy-controller -y vsphere --destroy-all-models
(ensure datastore cache is deleted automatically)

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1711019

### CODE REVIEW

File: provider/vsphere/internal/vsphereclient/createvm.go [hunk 1]
```diff
@@ -40,11 +35,22 @@ type CreateVirtualMachineParams struct {
...
+	// this controller. Within this directory there will be subdirectories
+	// for each series, and within those the VMDKs will be stored.
+	VMDKDirectory string
+
+	// Series is the name of the OS series that the image will run.
+	Series string
 
-	// OVF contains the OVF content.
-	OVF string
+	// ReadOVA returns the location of, and an io.ReadCloser for
```
Comment (wallyworld) [inline provider/vsphere/internal/vsphereclient/createvm.go]: needs comma at end of line

File: provider/vsphere/internal/vsphereclient/createvm.go [hunk 2]
```diff
@@ -133,78 +136,100 @@ func (c *Client) CreateVirtualMachine(
...
+	// After cloning the temporary VM, we must detach the original
+	// VMDK from the temporary VM to avoid deleting it when destroying
+	// the VM.
+	c.logger.Debugf("cloning VM")
+	vm, err := c.cloneVM(ctx, tempVM, args.Name, vmFolder, taskWaiter)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer func() {
+		if err == nil {
```
Comment (wallyworld) [inline provider/vsphere/internal/vsphereclient/createvm.go]: I can't quite follow what err this is referencing.
Shouldn't we use a named return error for the entire method and this err would then shadow that return error value?

File: provider/vsphere/internal/vsphereclient/status.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+type taskWaiter struct {
+	clock                  clock.Clock
+	updateProgress         func(string)
+	updateProgressInterval time.Duration
+}
+
+func (w *taskWaiter) WaitTask(ctx context.Context, t *object.Task, action string) (*types.TaskInfo, error) {
```
Comment (wallyworld) [inline provider/vsphere/internal/vsphereclient/status.go]: comment?

File: provider/vsphere/internal/vsphereclient/createvm.go (hunk 1 above)
Comment (axw) [reply provider/vsphere/internal/vsphereclient/createvm.go]: fixed

File: provider/vsphere/internal/vsphereclient/createvm.go (hunk 2 above)
Comment (axw) [reply provider/vsphere/internal/vsphereclient/createvm.go]: done

File: provider/vsphere/internal/vsphereclient/status.go (hunk 3 above)
Comment (axw) [reply provider/vsphere/internal/vsphereclient/status.go]: unexported it instead

**Comment by axw** [comment]
$$merge$$

**Pull request merged**